subprojects/plugin-use/src/main/java/org/gradle/plugin/use/internal/DefaultPluginRequestApplicator.java

```diff
@@ -85,63 +88,90 @@ public DefaultPluginRequestApplicator(
         final Map<Result, PluginImplementation<?>> pluginImplsFromOtherLoaders = newLinkedHashMap();
 
+        List<Result> results = resolvePluginRequests(requests, effectivePluginResolver);
         if (!results.isEmpty()) {
-            for (final Result result : results) {
-                applyPlugin(result.request, result.found.getPluginId(), new Runnable() {
-                    @Override
-                    public void run() {
-                        result.found.execute(new PluginResolveContext() {
-                            @Override
-                            public void addLegacy(PluginId pluginId, Object dependencyNotation) {
-                                pluginApplyActions.add(target -> applyLegacyPlugin(target, result, pluginId));
-                                scriptHandler.addScriptClassPathDependency(dependencyNotation);
-                            }
-
-                            @Override
-                            public void add(PluginImplementation<?> plugin) {
-                                pluginApplyActions.add(target -> applyPlugin(target, result, plugin));
-                            }
-
-                            @Override
-                            public void addFromDifferentLoader(PluginImplementation<?> plugin) {
-                                pluginApplyActions.add(target -> applyPlugin(target, result, plugin));
-                                pluginImplsFromOtherLoaders.put(result, plugin);
-                            }
-                        });
-                        String pluginVersion = result.found.getPluginVersion();
-                        if (pluginVersion != null) {
-                            pluginVersionTracker.setPluginVersionAt(
-                                classLoaderScope,
-                                result.found.getPluginId().getId(),
-                                pluginVersion
-                            );
-                        }
-                    }
-                });
-            }
+            applyPlugins(scriptHandler, classLoaderScope, pluginApplyActions, pluginImplsFromOtherLoaders, results);
         }
 
         defineScriptHandlerClassScope(scriptHandler, classLoaderScope, pluginImplsFromOtherLoaders.values());
+
+        if (!autoAppliedPlugins.isEmpty()) {
+            final PluginResolver alreadyOnClasspathIgnoringPluginResolver = new AlreadyOnClasspathIgnoringPluginResolver(effectivePluginResolver,
+                new ClassloaderBackedPluginDescriptorLocator(scriptHandler.getClassLoader())
+            );
+
+            List<Result> autoAppliedPluginResults = resolvePluginRequests(autoAppliedPlugins, alreadyOnClasspathIgnoringPluginResolver);
+            if (!autoAppliedPluginResults.isEmpty()) {
+                ((DefaultClassLoaderScope) classLoaderScope).dropClassloaders();
+                ((DefaultClassLoaderScope) classLoaderScope).unlock();
```

Comment:
This is not terrible, but it feels a little dirty.  It seems like we have a couple of choices here:
- We reset the classloader scope after checking if we need to add auto-applied plugins.  This is what this PR is currently doing.
- We change the classloader scope structure such that explicit plugins are resolved separately from auto-applied plugins - for instance, one is the parent of the other.  This might be tricky since the classloader scope that gets passed in here is the classloader scope for the domain object, so adding another layer might have consequences and introduce some complexity.
- We don't use the classloader scope and just construct the classloader that we scan for plugins without any caching (at this level, at least).  I believe this is what effectively happens if we reset the classloader scope - we use it once and then throw away the cached classloader.  The only difference is that we could potentially keep using it if we decide the auto-applied plugin is already on the classpath.  Since there is only one auto-applied plugin in existence currently (enterprise plugin) and it is applied to the settings object, I wonder how inefficient it would actually be to just construct an un-cached classloader for checking if the plugin is already applied?  Then we could not lock the classloader scope until we know exactly what will be in it.  We'd be taking on some technical debt if we ever had more auto-applied plugins, but I don't know when/if we ever will.

Any other ideas?

Reply:
What I want to have is that the usual use-case is optimized. I think there are two of them:
- Without auto-applied plugins (without `--scan`): We should do what we do now and only construct and lock the classloader scope once.
- With auto-applied plugins (with `--scan`), though with the scan plugin already applied: We should do what we do now and only construct and lock the classloader scope once, so we don't incur any performance overhead.

My current assumption is that you'd more often have the enterprise plugin already applied when you use `--scan`, given that publishing to scans.gradle.com is probably not happening so often and the CI injections scripts and any local init scripts apply the plugin already. In that sense, you'd use the `--scan` switch only to switch on publishing a scan and not for applying the enterprise plugin.

Though in the end, it doesn't probably matter very much as long as we keep the case when you don't have `--scan` as fast as it is now.

Reply:
I think that makes sense.  Do you think it would be better if, instead of exposing unlock/dropClassLoder methods, we have something more encapsulated, where we can better enforce that this is something we should only do while building the classloader scope?  Something like the following on DefaultClassLoaderScope:

```
    public void lockWith(ModifyingLockActor lockActor) {
        if (locked) {
            throw new IllegalStateException("class loader scope is already locked, scope identifier is " + id)
        } else {
            lock();
        }

        if (lockActor.requiresModification()) {
            dropClassloaders();
            unlock();
            lockActor.modify(this);
            lock();
        }
    }
    
    interface ModifyingLockActor {
        boolean requiresModification();
        void modify(ClassLoaderScope scope);
    }
```

And then in `PluginRequestApplicator` we do something like:

```
classLoaderScope.lockWith(new ClassLoaderScope.ModifyingLockActor() {
                private List<Result> autoAppliedPluginRequests;
                
                @Override
                public boolean requiresModification() {
                    autoAppliedPluginRequests = resolvePluginRequests(autoAppliedPlugins, alreadyOnClasspathIgnoringPluginResolver);
                    return !autoAppliedPluginRequests.isEmpty();
                }

                @Override
                public void modify(ClassLoaderScope scope) {
                    scriptHandler.dropResolvedClassPath();

                    final Map<Result, PluginImplementation<?>> pluginImplementations = newLinkedHashMap();
                    applyPlugins(scriptHandler, classLoaderScope, pluginApplyActions, pluginImplementations, autoAppliedPluginRequests);
                    defineScriptHandlerClassScope(scriptHandler, classLoaderScope, pluginImplementations.values());
                }
            });
```
 I wouldn't want to introduce something that allows someone to just arbitrarily unlock a classloader scope at any point.

Reply:
Yes, I think that would be better. I would prefer to throw away the classloader scope instance, or make a mutable copy of it if we need to modify it instead of trying to unlock things. Though given how the code is currently structured that would make it more complex.

Reply:
@wolfs @ghale I think we can avoid all this **unlocking/locking/throwing stuff away** without introducing too much complexity. 

```
public class MutableClassLoaderScope implements ClassLoaderScope {

    private ClassLoaderScope delegate;

    public MutableClassLoaderScope(
        ClassLoaderScope delegate
    ) {
        this.delegate = delegate;
    }

   ...

    @Override
    public ClassLoaderScopeId getId() {
        return delegate.getId();
    }

   ...
 
    public void mutate(MutateActor mutateActor) {
        if (mutateActor.requiresMutation()) {
            ClassLoaderScope child = delegate.createChild(delegate.getId() + "-child", null);
            mutateActor.modify(child);
            child.lock();

            delegate = child;
        }
    }


    public interface MutateActor {
        boolean requiresMutation();

        void modify(ClassLoaderScope scope);
    }
}
```

And we would use it like this:

```
        classLoaderScope.mutate(new MutableClassLoaderScope.MutateActor() {
            private List<Result> autoAppliedPluginRequests;

            @Override
            public boolean requiresMutation() {
                final PluginResolver alreadyOnClasspathIgnoringPluginResolver = new AlreadyOnClasspathIgnoringPluginResolver(effectivePluginResolver,
                    new ClassloaderBackedPluginDescriptorLocator(scriptHandler.getClassLoader())
                );

                autoAppliedPluginRequests = resolvePluginRequests(autoAppliedPlugins, alreadyOnClasspathIgnoringPluginResolver);

                return !autoAppliedPluginRequests.isEmpty();
            }

            @Override
            public void modify(ClassLoaderScope scope) {
                scriptHandler.dropResolvedClassPath();

                final Map<Result, PluginImplementation<?>> pluginImplementations = newLinkedHashMap();
                applyPlugins(scriptHandler, classLoaderScope, pluginApplyActions, pluginImplementations, autoAppliedPluginRequests);
                defineScriptHandlerClassScope(scriptHandler, classLoaderScope, pluginImplementations.values());
            }
        });
```

And [wrap](https://github.com/gradle/gradle/blob/ea1f76efda73a4789114a338c0b7719fdbdf7ef8/subprojects/core/src/main/java/org/gradle/api/internal/initialization/AbstractClassLoaderScope.java#L88) the `DefaultClassLoaderScope` in a `MutableClassLoaderScope`.

Even better, we can check if the delegate is locked and either create a child or mutate the delegate based on that.

Reply:
I don't think this is a bad idea. Though I wonder if this would confuse some other code since the classloader scope will now suddenly have a different name. Let's see what @ghale thinks about this.

Reply:
I like it, too.  I think we'll want to be very careful how this gets handled when serializing/unserializing from the configuration cache, and I think we want to use a descriptive name for the child scope (e.g. "-plugins" or something like that) but otherwise I like the idea that it keeps everything tidy inside the scope itself.